controllers/registry_controller.go

```diff
@@ -71,9 +85,118 @@ func (r *registryReconciler) Reconcile(ctx context.Context, req ctrl.Request) (r
 		err = markAsFreeOnSuccess(r.client, r.watchNamespace, r.recorder, RegistryController, result.Requeue, err)
 	}()
 
+	// List all IDMS/ITMS resources
+	imageDigestMirrorSetList := &config.ImageDigestMirrorSetList{}
+	if err = r.client.List(ctx, imageDigestMirrorSetList); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting IDMS list: %w", err)
+	}
+	imageTagMirrorSetList := &config.ImageTagMirrorSetList{}
+	if err = r.client.List(ctx, imageTagMirrorSetList); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error getting ITMS list: %w", err)
+	}
+
+	mergedMirrorSets := collectMirrorSets(imageDigestMirrorSetList.Items, imageTagMirrorSetList.Items)
+
 	return ctrl.Result{}, nil
 }
 
+// collectMirrorSets extracts the contents of the given mirror sets and merges them into a single collection.
+// Each member of the collection represents the registry configuration for a specific source.
+func collectMirrorSets(idmsItems []config.ImageDigestMirrorSet, idtsItems []config.ImageTagMirrorSet) []mirrorSet {
+	var allMirrorSets []mirrorSet
+	for _, idms := range idmsItems {
+		for _, entry := range idms.Spec.ImageDigestMirrors {
+			set := &mirrorSet{
+				source:             entry.Source,
+				mirrorSourcePolicy: entry.MirrorSourcePolicy,
+			}
+			for _, image := range entry.Mirrors {
+				set.mirrors = append(set.mirrors, mirror{host: string(image), resolveTags: false})
+			}
+			allMirrorSets = append(allMirrorSets, *set)
+		}
+	}
+	for _, itms := range idtsItems {
+		for _, entry := range itms.Spec.ImageTagMirrors {
+			set := &mirrorSet{
+				source:             entry.Source,
+				mirrorSourcePolicy: entry.MirrorSourcePolicy,
+			}
+			for _, image := range entry.Mirrors {
+				set.mirrors = append(set.mirrors, mirror{host: string(image), resolveTags: true})
+			}
+			allMirrorSets = append(allMirrorSets, *set)
+		}
+	}
```

Comment:
Wouldn't it be easier to merge the mirrorsets as you are creating the initial list?

start with a map,
check if each new mirrorset already is in the map
if not -> add to map
if it is -> add mirror to map

Reply:
I actually tried this earlier but it made it difficult to unit test the merging logic